### ISSUE TITLE
`DefaultDependencies=no` invokes two different behaviors that most units don't want simultaneously

### ISSUE THREAD

**Issue by joshtriplett** [systemd/systemd#10210]
**systemd version the issue has been seen with**
> v239

**Used distribution**
 > Distro-independent; based on observations made from both Debian and NixOS.

`DefaultDependencies=no` in a systemd unit has two completely independent behaviors:

1) It makes the unit not gain the default "dependencies of type `Requires=` and `After=` on `sysinit.target`, a dependency of type `After=` on `basic.target` as well as dependencies of type `Conflicts=` and `Before=` on `shutdown.target`.

2) It makes any `.target` that declares a `Wants` or `Requires` on the unit (or equivalently has `.wants` or `.requires` directory entries, etc) *not* also have an `After` on that unit. (Note that this behavior applies whether either the `.target` has `DefaultDependencies=no` *or* the unit it declares a `Wants` or `Requires` on has `DefaultDependencies=no`.)

This has several issues.

(Let's leave aside for the moment that it isn't at all obvious when you'd ever *want* to have a `Wants` or `Requires` from a `.target` on a unit but not have an `After` on that same unit; that seems like it somewhat defeats the purpose of having a `.target`. In general, I can see the use case for having `After` without `Wants`, but not `Wants` without `After`.)

As far as I can tell, there's no correlation between services wanting one of those two behaviors and wanting the other.

For example, every early-boot service in systemd that uses `DefaultDependencies=no` for behavior (1) proceeds to *re-add* a `Before=sysinit.target` (or `Before` some other unit that in turn has a `Before` on `sysinit.target`), undoing behavior (2).

Likewise, a `.target` that declares `DefaultDependencies=no` also seems likely to want to control these behaviors independently; for instance, as far as I can tell the various stock `.target` files in systemd that declare `DefaultDependencies=no` want behavior (1), not behavior (2).

Conversely, if someone *did* have a use case for behavior (2) of `DefaultDependencies=no` (omitting the implicit `After` from any `.target` that has a `Wants` or `Requires` on the unit), they're unlikely to also want behavior (1) of `DefaultDependencies=no`, so they'd have to re-declare the default dependencies on `sysinit.target`, `basic.target`, and `shutdown.target`.

And there's no clear link in the documentation for people who use `DefaultDependencies=no` for behavior (1) to realize that it *also* invokes behavior (2), which seems highly likely to result in surprising behavior for people trying to construct targets. (Doubly so because the mentions of `DefaultDependencies=no` in the documentation that discuss behavior (1) do not also cross-reference behavior (2), making it even less likely that people would know it invokes both behaviors.)

Given this, I would suggest two things:

First, could the relevant manpages that currently discuss these behaviors clearly document the use case for behavior (2)? It isn't at all obvious when you would want that behavior for the dependencies of a `.target`, and I think *not* doing this would make many existing dependency graphs much clearer and more robust.

And second, if behavior (2) does have a concrete use case, then could these two behaviors get separated into separate directives? That would simplify the numerous (AFAICT majority) cases that want only one or the other behavior by not making them subsequently have to explicitly declare dependencies to undo the behavior they don't want. (This would need a careful read-through of users of `DefaultDependencies=no` to figure out if any of them actually *want* behavior (2), perhaps with a look at the dependency graph.)

**Comment by jameysharp** [comment]
Thanks for filing this, Josh. I spent a good portion of my afternoon today being baffled by this combination of behaviors, so at the very least I too would love to see better documentation. Even better if these two behaviors start being controlled by separate options, assuming there really is a use case for behavior (2).

**Comment by Werkov** [comment]
Hi. I agree there are the two distinct manifestations of `DefaultDependencies=` (AFAIK there are even more, one per each unit type). However, for me the [explanation](https://github.com/systemd/systemd/blob/master/man/systemd.unit.xml#L259) in `man systemd.unit` and related man pages provides the complete image. What man pages in what way would you suggest to modify?

P.S. I think a source of some confusion might be substitution of "service" with "unit" in your definition of behavior (1).

**Comment by jameysharp** [comment]
You're right that Josh described behavior (1) with respect to services specifically, but as far as I can tell the cause of confusion here applies to all unit types; our concern is really about units' interactions with targets.

In my opinion, the documentation problem is that `systemd.target` is the only place to discover the effect that `DefaultDependencies=no` in _other_ units has on a target that declares a `Wants` or `Requires` relationship to those units.

In all other cases, `DefaultDependencies` only has an effect on dependencies which systemd adds for that unit itself. Targets are weird because _either_ end of the `Wants`/`Requires` dependency can disable the target's default dependencies.

To describe the current behavior clearly, I think all the other unit types should mention that by default they have a `Before` relationship with any `.target` for which they have a `WantedBy` or `RequiredBy` relationship, unless the target has `DefaultDependencies=no`. Otherwise they give an incomplete impression of the effect that the `DefaultDependencies` option has.

Going further, I think the fact that this behavior is remarkably difficult to explain is a mark against it, and I'd love to hear why it exists.

**Comment by bl33pbl0p** [comment]
@joshtriplett 
> In general, I can see the use case for having After without Wants, but not Wants without After.)

I'd say it is rather After= alone that is difficult to explain for something, unless you are constructing it in a setup where it is pulled by something else (anchoring the transaction) so you just care about ordering when these set of jobs are pulled in together.

Wants= is very much a valid use case \*without\* After=, as that allows you to trigger a start job for it, and then just move on: not wait on it, not be failed by it (doesn't happen with Wants=, but you get the idea). It also makes sense for a target, as then you can use it as some sort of anchor to pull in things and still not care about waiting for all to come up.

The thing about DefaultDependencies= is that in systemd "dependencies" include propagation, ordering, and triggers. Therefore, After=/Before= are also dependencies, just like Wants=/BindsTo=/Requires=/PartOf=. They are ordering ones, the others are requirement/propagation dependencies. DefaultDependencies=no hence simply disables everything something gains by default.

@jameysharp 
> In my opinion, the documentation problem is that systemd.target is the only place to discover the effect that DefaultDependencies=no in other units has on a target that declares a Wants or Requires relationship to those units.

I don't follow. How would having DD=no in something else have any effect on a target it references?

> To describe the current behavior clearly, I think all the other unit types should mention that by default they have a Before relationship with any .target for which they have a WantedBy or RequiredBy relationship, unless the target has DefaultDependencies=no. Otherwise they give an incomplete impression of the effect that the DefaultDependencies option has.

It is instead a symmetric automatic property they gain because the thing that they are WantedBy= ends up creating a .wants directory, which is Wants= for it and thus has an implicit After= (by virtue of being a target). I think this is clear enough already.

**Comment by boucman** [comment]
maybe better document that WantedBy and RequiredBy can add default dependecies du to the nature of the other end of the relationship?

**Comment by jameysharp** [comment]
@bl33pbl0p I think you've agreed with me:

> I don't follow. How would having DD=no in something else have any effect on a target it references?

Per `systemd.target`,

> The following dependencies are added unless DefaultDependencies=no is set:
>
> Target units will automatically complement all configured dependencies of type Wants= or Requires= with dependencies of type After= unless DefaultDependencies=no is set in the specified units.

So you have to check for DefaultDependencies both in the target and in every unit that it Wants (which is nontrivial to determine, due to WantedBy and .wants) in order to predict what dependencies systemd will generate. This is even more confusing than the behavior of default dependencies for other types of units.

**Comment by bl33pbl0p** [comment]
For the record, this is not limited to targets, this happens with all unit types.

**Comment by bl33pbl0p** [comment]
So to summarise, the WHY behind this whole special check is probably this:

When you create a reverse dependency WantedBy=, the service wants to expect that it loses all default dependencies when it sets DD=no, but targets .wants/Wants= created in this case are special, as until the *target* as DD=no, the ordering dep is kept unconditionally. So this is the reason behind why this was done, because targets are kind of special.

Let's evaluate what happens if we change it to be more uniform like other units, now when you set DD=no, you still get ordering dep on the target (Before= by default). This means now you must either set DD=no in target to lose it, and then specify After= manually for all other units, or specifically modify target to either set Before=service (which might be something you don't want, you may not want ordering). This ends up being too much work.

Now, the correct way to solve this would be to see which case is more common. Either solution makes one case hard and the other easy, so if with the patch things become more easy for the common use case, then we should be opting for that. Also, not sure if this would be a compat breakage.

@jameysharp wdyt?

**Comment by joshtriplett** [comment]
On Sun, Jan 20, 2019 at 11:03:00AM -0800, bl33pbl0p wrote:
> Now, the correct way to solve this would be to see which case is more common. Either solution makes one case hard and the other easy, so if with the patch things become more easy for the common use case, then we should be opting for that. Also, not sure if this would be a compat breakage.

The other question to ask is which behavior is less surprising.

**Comment by tmccombs** [comment]
It looks like possibly a way to get the (2) behavior without (1), sort of, is to explicitly add an `After=` dependency on the target.  The downside of course is that then it can't be started in parallel with other services in the target.

And a use case for (2) is if you want to automatically start a service, but don't want to wait for it to finish for the target to complete. I ran into this where I had a service that didn't start do to an ordering cycle because it depended on a service with an After= dependency on the target that I was installing the service into with a WantedBy=. It took me a while to figure out where the Before= dependency was coming from, because I didn't think to look at the systemd.target manpage. 

I wonder if maybe there is some way that systemd could allow opting out of a single default dependency instead of all of them, for example, maybe something like:
```
Before=-multi-user.target
```
which means roughly "if there is a Before=multi-user.target dependency added by a default dependency rule, then ignore it".  That would also allow you to do something like remove the default depenencies on basic.target without removing the dependencies on sysinit.target.
